### PR TITLE
ISPN-10249 Upgrade to Hibernate Search 6

### PR DESCRIPTION
**Pull request by fax4ever**
https://issues.redhat.com/browse/ISPN-10249

Reopening of https://github.com/infinispan/infinispan/pull/8226

### CODE REVIEW

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/event/ClientListenerWithDslFilterObjectStorageTest.java [hunk 1]
```diff
@@ -81,8 +81,7 @@ protected ConfigurationBuilder getConfigurationBuilder() {
       cfgBuilder.encoding().key().mediaType(MediaType.APPLICATION_OBJECT_TYPE);
       cfgBuilder.encoding().value().mediaType(MediaType.APPLICATION_OBJECT_TYPE);
       cfgBuilder.indexing().enable()
-            .addProperty("default.directory_provider", "local-heap")
-            .addProperty("lucene_version", "LUCENE_CURRENT");
```
Comment (yrodiere) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/event/ClientListenerWithDslFilterObjectStorageTest.java:85]: I think you don't have to remove this, just replace `LUCENE_CURRENT` with `LATEST`.

Comment (yrodiere) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/event/ClientListenerWithDslFilterObjectStorageTest.java:85]: Though this property is indeed not necessary (LATEST is the default)

File: query/src/test/java/org/infinispan/query/dsl/embedded/impl/HibernateSearchPropertyHelperTest.java [hunk 2]
```diff
@@ -20,14 +21,14 @@
...
-   public SearchFactoryHolder factoryHolder = new SearchFactoryHolder(TestEntity.class);
-
    private HibernateSearchPropertyHelper propertyHelper;
 
    @Before
    public void setup() {
-      propertyHelper = new HibernateSearchPropertyHelper(factoryHolder.getSearchFactory(), new ReflectionEntityNamesResolver(null));
+      SearchMappingHolder mappingHolder = mock(SearchMappingHolder.class);
+      when(mappingHolder.getSearchMapping()).thenReturn(
+            SearchMappingHelper.createSearchMappingForTests(TestEntity.class));
```
Comment (yrodiere) [inline query/src/test/java/org/infinispan/query/dsl/embedded/impl/HibernateSearchPropertyHelperTest.java:30]: @fax4ever Unless I'm mistaken, you're not closing the search mapping, which leads to the test failures in CI mentioning "leaked threads".

There are other test failures mentioning thread leaks, maybe you should have a look at them? The name of the test generally appears somewhere in the stack trace.

Comment (fax4ever) [reply query/src/test/java/org/infinispan/query/dsl/embedded/impl/HibernateSearchPropertyHelperTest.java:30]: Sure. They seems to be related to `LuceneTransformationTest` test instance.

Comment (fax4ever) [reply query/src/test/java/org/infinispan/query/dsl/embedded/impl/HibernateSearchPropertyHelperTest.java:30]: in this test we open and (close -- maybe not) a search mapping instance manually to check the Lucene queries derived from it

Comment (yrodiere) [reply query/src/test/java/org/infinispan/query/dsl/embedded/impl/HibernateSearchPropertyHelperTest.java:30]: > Sure. They seems to be related to LuceneTransformationTest test instance.

Actually they are not. If you look at the stack trace, the threads were created in other tests. `HibernateSearchPropertyHelperTest` in particular, but there may be others.

File: query/src/main/java/org/infinispan/query/clustered/ClusteredQueryOperation.java [hunk 3]
```diff
@@ -38,12 +38,12 @@ public QueryDefinition getQueryDefinition() {
       return queryDefinition;
    }
 
-   static ClusteredQueryOperation createLazyIterator(QueryDefinition queryDefinition, UUID queryId) {
+   /*static ClusteredQueryOperation createLazyIterator(QueryDefinition queryDefinition, UUID queryId) {
```
Comment (gustavonalle) [inline query/src/main/java/org/infinispan/query/clustered/ClusteredQueryOperation.java:41]: heads up, we've deprecated the whole "CacheQuery" class in 11, and this include the so called "lazy" iterator. So you could just get rid of it

File: query/src/main/java/org/infinispan/query/clustered/commandworkers/package-info.java [hunk 4]
```diff
@@ -1,4 +0,0 @@
-/**
- * @private
- */
-package org.infinispan.query.clustered.commandworkers;
```
Comment (gustavonalle) [inline query/src/main/java/org/infinispan/query/clustered/commandworkers/package-info.java:4]: Shouldn't be deleted

File: query/src/main/java/org/infinispan/query/dsl/embedded/impl/EmbeddedLuceneQuery.java [hunk 5]
```diff
@@ -83,21 +83,23 @@ public void resetQuery() {
...
    public CloseableIterator<T> iterator() {
-      CloseableIterator<T> iterator = createCacheQuery().iterator();
-      return rowProcessor == null ? iterator : new MappingIterator(iterator, t -> rowProcessor.apply((Object[]) t));
+      IndexedQuery<T> cacheQuery = createCacheQuery();
+      return new MappingIterator(cacheQuery.iterator(), i -> (projection == null) ? i : convertProjectionItem(i) );
    }
 
    @Override
    public int getResultSize() {
-      //todo [anistor] optimize this by running a slightly modified query that performs just COUNT only, ignoring projections or sorting
```
Comment (gustavonalle) [inline query/src/main/java/org/infinispan/query/dsl/embedded/impl/EmbeddedLuceneQuery.java:99]: Did you fix this @fax4ever ? If not, let's keep it as a reminder :)

File: query/src/main/java/org/infinispan/query/impl/CacheQueryImpl.java [hunk 6]
```diff
@@ -42,83 +36,36 @@
...
-    */
-   public CacheQueryImpl(Query luceneQuery, SearchIntegrator searchFactory, AdvancedCache<?, ?> cache,
-                         KeyTransformationHandler keyTransformationHandler, TimeoutExceptionFactory timeoutExceptionFactory,
-                         Class<?> entity) {
-      this(timeoutExceptionFactory == null ? searchFactory.createHSQuery(luceneQuery, entity) :
-                  searchFactory.createHSQuery(luceneQuery, entity).timeoutExceptionFactory(timeoutExceptionFactory),
-            cache, keyTransformationHandler);
-   }
+   // caching result size, since Search 6 doesn't do that
+   private Long resultSize;
```
Comment (gustavonalle) [inline query/src/main/java/org/infinispan/query/impl/CacheQueryImpl.java:43]: We are not caching query result size or list anymore:  https://infinispan.org/docs/stable/titles/upgrading/upgrading.html#querying 
Is this something you need to do?

File: build-configuration/pom.xml [hunk 7]
```diff
@@ -2,12 +2,6 @@
 <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>
 
-   <parent>
-      <groupId>org.jboss</groupId>
-      <artifactId>jboss-parent</artifactId>
-      <version>36</version>
-   </parent>
```
[the remark this answers is not in the document]
Comment (pruivo) [reply build-configuration/pom.xml:9]: The `jboss-parent` is mandatory downstream. Please keep it upstream so we can detect issues sooner than later.

File: query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java [hunk 8]
```diff
@@ -322,69 +314,66 @@ public Object visitReadWriteManyEntriesCommand(InvocationContext ctx, ReadWriteM
...
+      SearchMapping searchMapping = searchMappingHolder.getSearchMapping();
+      if (searchMapping == null) {
+         return;
+      }
+
+      searchMapping.scope(entityType).workspace().purge();
+   }
+
+   public void refreshIndex(Class<?> entityType) {
+      SearchMapping searchMapping = searchMappingHolder.getSearchMapping();
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java:335]: Refresh will create new index readers. This is mostly only relevant if you tell Hibernate Search to reuse out-of-date index readers by setting `hibernate.search.backends.myBackend.index_defaults.io.refresh_interval` to something higher than 0.

See https://docs.jboss.org/hibernate/search/6.0/reference/en-US/html_single/#backend-lucene-io-refresh

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/RemoteQueryDslConditionsTunedTest.java [hunk 9]
```diff
@@ -29,23 +30,26 @@ protected ConfigurationBuilder getConfigurationBuilder() {
...
-             .addProperty("default.locking_strategy", "native")
-             .addProperty("default.sharding_strategy.nbr_of_shards", String.valueOf(NUM_SHARDS));
+             .addProperty(SearchConfig.DIRECTORY_TYPE, SearchConfig.FILE)
+             .addProperty(SearchConfig.DIRECTORY_ROOT, indexDirectory)
+             .addProperty(SearchConfig.THREAD_POOL_SIZE, String.valueOf(NUM_SHARDS))
+             .addProperty(SearchConfig.QUEUE_COUNT, String.valueOf(NUM_SHARDS))
+             .addProperty(SearchConfig.QUEUE_SIZE, "4096")
+             .addProperty(SearchConfig.COMMIT_INTERVAL, "10000")
+             .addProperty(SearchConfig.SHARDING_STRATEGY, SearchConfig.HASH)
+             .addProperty(SearchConfig.NUMBER_OF_SHARDS, String.valueOf(NUM_SHARDS));
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/RemoteQueryDslConditionsTunedTest.java:40]: I think we discussed this already a few months ago, but let's summarize it here.

NRT is enabled by default, sort of. Commits are asynchronous by default and index readers are aware of uncommitted changes by default.

Configuration in Hibernate Search 6 is different from Hibernate Search 5:

* When you request indexing, you (Infinispan) can request that the commit happens immediately.
  * If you do, the commit will happen before the indexing operation returns. Obviously, it will perform badly if you index your documents one by one.
  * If you don't the commit will happen some time after the indexing operation returns. By default it's 1000ms, but you can set it to a different number.

Readers are always aware of uncommitted changes by default, i.e. they are aware of changes still in the index writer buffers.

If you can accept out-of-date readers, you can set a refresh interval so that readers are only re-created every 1000ms, for example.

More information: https://docs.jboss.org/hibernate/search/6.0/reference/en-US/html_single/#backend-lucene-io

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/admin/RemoteCacheAdminTest.java [hunk 10]
```diff
@@ -46,7 +47,7 @@ protected void createCacheManagers() throws Throwable {
             getDefaultClusteredCacheConfig(CacheMode.DIST_SYNC, false));
       builder.indexing().enable()
              .addIndexedEntity("sample_bank_account.Transaction")
-             .addProperty("default.directory_provider", "local-heap");
```
[the remark this answers is not in the document]
Comment (yrodiere) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/admin/RemoteCacheAdminTest.java:49]: You can transform the properties before passing them to Hibernate Search, of course. But there is no plan to handle that kind of things in Hibernate Search 6.

In fact, we even issue a warning when someone uses old properties.

You can find a list of old properties here, if that's useful: https://github.com/hibernate/hibernate-search/blob/0220c7a345ba46f9559e73ec5ae5d9ccddbec2df/engine/src/main/java/org/hibernate/search/engine/cfg/spi/HibernateSearch5Properties.java

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/RemoteQueryDslConditionsTunedTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (yrodiere) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/RemoteQueryDslConditionsTunedTest.java:40]: > is that still the case in Hibernate Search 6?

If you force the commit, it will happen ASAP, so no.

If you don't force the commit, then Hibernate Search will still return before the commit. However, if you set commit interval is set to 0, it will commit just after each "batch" of operations. The size of batches can be configured.

If you want guarantees that changes are flushed to disk, that's obviously not enough. Depending on Infinispan's needs, we could introduce another behavior where you don't force the commit, but ask Hibernate Search to only consider the indexing operation completed when the next commit happens (whenever that is). I believe that shouldn't be too much work. That will have to wait until after this PR is merged, though (we need another release of Search).

[the remark this answers is not in the document]
Comment (gustavonalle) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/RemoteQueryDslConditionsTunedTest.java:40]: Hold on a sec. So basically, we are moving from

1) ```master```, where NRT is not enabled by default, and will commit on every document synchronously, but will coalesce writing threads to reduce the number of commits. Once the cache operation completes, it is guarantee the document is indexed on disk and it's visible on searches.

to

2) ```ISPN-10249``` (this PR), where NRT is ON by default, commits are done by batch size or time, cache operations guarantee immediate visibility on searches but not immediate persistence of the index, but only once every second (configurable)

Correct?

If so, I reckon 2) is overall better than 1), so I wouldn't rush implementing the old behaviour

[the remark this answers is not in the document]
Comment (yrodiere) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/RemoteQueryDslConditionsTunedTest.java:40]: That's

[the remark this answers is not in the document]
Comment (yrodiere) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/RemoteQueryDslConditionsTunedTest.java:40]: Correct.

The only drawback of the new behavior is there is no way to get a guarantee that each change is on disk after the cache operation completes unless you accept bad performance (one commit per cache operation).

I believe that kind of requirement is a bit unreasonable for a secondary source of truth such as Lucene indexes, though. We will probably have a look at some time in the future, for the sake of completeness, but I agree with you this can wait.

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/admin/RemoteCacheAdminTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (fax4ever) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/admin/RemoteCacheAdminTest.java:49]: As @yrodiere said there is no one-to-one correspondence for some properties.
For those for which there is, we could create such a compatibility layer in `org.infinispan.search.mapper.mapping.impl.IndexProperties`.
At the moment in this class we simplify a little the prefix needed to be used by the user to configure Hibernate Search properties, so it looks like the right place to implement it.

File: query/src/main/java/org/infinispan/query/spi/SearchManagerImplementor.java [hunk 11]
```diff
@@ -15,33 +12,13 @@
...
-   @Deprecated
-   void setTimeoutExceptionFactory(TimeoutExceptionFactory timeoutExceptionFactory);
-
    /**
     * Creates a cache query based on a {@link QueryDefinition} and a custom metadata.
     * @deprecated Since 11.0, without replacement. To be removed in next major version.
     */
    @Deprecated
-   <E> CacheQuery<E> getQuery(QueryDefinition queryDefinition, IndexedQueryMode indexedQueryMode, IndexedTypeMap<CustomTypeMetadata> indexedTypeMap);
+   <E> CacheQuery<E> getQuery(QueryDefinition queryDefinition, IndexedQueryMode indexedQueryMode);
```
[the remark this answers is not in the document]
Comment (fax4ever) [reply query/src/main/java/org/infinispan/query/spi/SearchManagerImplementor.java:20]: Sure

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/event/ClientListenerWithDslFilterObjectStorageTest.java (hunk 1 above)
Comment (fax4ever) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/event/ClientListenerWithDslFilterObjectStorageTest.java:85]: Shall I replace the old `LUCENE_CURRENT` with the `LATEST`? I removed them since in general I prefer have less code than more :P and in this case `LATEST` is the default one

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/admin/RemoteCacheAdminTest.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (gustavonalle) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/admin/RemoteCacheAdminTest.java:49]: Cool, we can defer this to a follow up PR

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/event/ClientListenerWithDslFilterObjectStorageTest.java (hunk 1 above)
Comment (yrodiere) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/event/ClientListenerWithDslFilterObjectStorageTest.java:85]: I think it was there to avoid the warning, maybe? I just mentioned this because I was surprised by this seemingly unnecessary change.

Anyway, it's probably not worth your time. Let's dropt his, sorry for the noise.

File: build-configuration/pom.xml (hunk 7 above)
[the remark this answers is not in the document]
Comment (fax4ever) [reply build-configuration/pom.xml:9]: I'm sorry. I had `http://repository.jboss.org/nexus/content/groups/public/` instead of `https://repository.jboss.org/nexus/content/groups/public-jboss/ ` as `jboss-public-repository-group` url in settings.xml. Now it works. I'll drop this commit.

File: query/src/main/java/org/infinispan/query/dsl/embedded/impl/EmbeddedLuceneQuery.java (hunk 5 above)
Comment (fax4ever) [reply query/src/main/java/org/infinispan/query/dsl/embedded/impl/EmbeddedLuceneQuery.java:99]: I have to check :P

File: build-configuration/pom.xml (hunk 7 above)
[the remark this answers is not in the document]
Comment (pruivo) [reply build-configuration/pom.xml:9]: for the record, you can add the option `-Dinsecure.repositories=WARN` to your `mvn` command if you had to use an insecure maven repository (i.e. non https).

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/event/ClientListenerWithDslFilterObjectStorageTest.java (hunk 1 above)
Comment (fax4ever) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/event/ClientListenerWithDslFilterObjectStorageTest.java:85]: @yrodiere No noise at all, thanks. I asked myself the same question when I did this change.

File: query/src/main/java/org/infinispan/query/spi/SearchManagerImplementor.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (fax4ever) [reply query/src/main/java/org/infinispan/query/spi/SearchManagerImplementor.java:20]: Unfortunately `IndexedTypeMap` is a Hibernate Search 5’s class. So I think we need to remove it. Doing now...

**Comment by anistor** [comment]
This needs rebasing again @fax4ever

File: core/src/main/java/org/infinispan/configuration/cache/IndexingConfigurationBuilder.java [hunk 12]
```diff
@@ -320,9 +320,9 @@ public IndexingConfiguration create() {
       // check for presence of index providers that are not persistent upon restart
       boolean isVolatile = typedProperties.entrySet().stream()
                                      .anyMatch(e -> {
-                                        if (((String) e.getKey()).endsWith(DIRECTORY_PROVIDER_SUFFIX)) {
+                                        if (((String) e.getKey()).endsWith("directory.type")) {
                                            String directoryImplementationName = String.valueOf(e.getValue()).trim();
-                                           return LOCAL_HEAP_DIRECTORY_PROVIDER.equalsIgnoreCase(directoryImplementationName)
+                                           return "local-heap".equalsIgnoreCase(directoryImplementationName)
```
Comment (anistor) [inline core/src/main/java/org/infinispan/configuration/cache/IndexingConfigurationBuilder.java:325]: Why this change ? LOCAL_HEAP_DIRECTORY_PROVIDER is "local-heap"

File: core/src/main/java/org/infinispan/configuration/cache/IndexingConfigurationBuilder.java [hunk 13]
```diff
@@ -320,9 +320,9 @@ public IndexingConfiguration create() {
       // check for presence of index providers that are not persistent upon restart
       boolean isVolatile = typedProperties.entrySet().stream()
                                      .anyMatch(e -> {
-                                        if (((String) e.getKey()).endsWith(DIRECTORY_PROVIDER_SUFFIX)) {
+                                        if (((String) e.getKey()).endsWith("directory.type")) {
```
Comment (anistor) [inline core/src/main/java/org/infinispan/configuration/cache/IndexingConfigurationBuilder.java:323]: DIRECTORY_PROVIDER_SUFFIX should probably become "directory.type". It is used in other places in this class, which I think are broken now unless we make the change

File: core/src/main/java/org/infinispan/configuration/cache/IndexingConfigurationBuilder.java (hunk 12 above)
Comment (fax4ever) [reply core/src/main/java/org/infinispan/configuration/cache/IndexingConfigurationBuilder.java:325]: Ideally I would have not wanted to change the core at all. This is (and the subsequent comment is about this same change) the only change we did to the ISPN core. The reason is that `isVolatile` must be true if and only if the local heap is used as directory type and we have different names in Search 6 for such property (key and value). I don't know actually if there is any other way to reach the same goal.

Comment (fax4ever) [reply core/src/main/java/org/infinispan/configuration/cache/IndexingConfigurationBuilder.java:325]: Sorry I didn't see well the constant value. Yes I think we can use it. I'm changing it...

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/RemoteQueryWithProtostreamAnnotationsTest.java [hunk 14]
```diff
@@ -31,7 +31,9 @@
  *
  * @author Adrian Nistor
  */
-@Test(testName = "client.hotrod.query.RemoteQueryWithProtostreamAnnotationsTest", groups = "functional")
+// TODO Hibernate Search 6 cannot handle more than one change of serialization context
```
[the remark this answers is not in the document]
Comment (fax4ever) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/RemoteQueryWithProtostreamAnnotationsTest.java:34]: @gustavonalle now that lifecycle has been changed so that all the protobuf entities are in the serialization context when the cache is starting, we don't need the skip and this comment anymore. Removing it now the test passes, so I'm removing it.

File: query/src/main/java/org/infinispan/query/impl/CacheQueryImpl.java (hunk 6 above)
Comment (fax4ever) [reply query/src/main/java/org/infinispan/query/impl/CacheQueryImpl.java:43]: Right! There is no reason to `emulate` a deprecated behavior in a deprecated API. I'm removing it.

File: query/src/main/java/org/infinispan/query/dsl/embedded/impl/EmbeddedLuceneQuery.java (hunk 5 above)
Comment (fax4ever) [reply query/src/main/java/org/infinispan/query/dsl/embedded/impl/EmbeddedLuceneQuery.java:99]: From what I can see it seems that we did it. Because `fetchTotalHitCount` Search 6 API invokes a specific work that does not involve pagination and sorting. @yrodiere Did I say something wrong? (Thanks)

Comment (yrodiere) [reply query/src/main/java/org/infinispan/query/dsl/embedded/impl/EmbeddedLuceneQuery.java:99]: You're right. And from what I can see `getResultSize()` calls `fetchTotalHitCount`, so we already optimize everything we need.

File: query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (fax4ever) [reply query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java]: Yes the method was in the wrong place, I removed it.

File: query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java [hunk 15]
```diff
@@ -322,69 +314,66 @@ public Object visitReadWriteManyEntriesCommand(InvocationContext ctx, ReadWriteM
...
-      for (int segment : segments) {
-         DeletionQuery deletionQuery = new SingularTermDeletionQuery(SEGMENT_FIELD, String.valueOf(segment));
-         for (IndexedTypeIdentifier type : searchFactory.getIndexBindings().keySet()) {
-            Work deleteWork = new DeleteByQueryWork(type, deletionQuery);
-            performSearchWork(deleteWork, NoTransactionContext.INSTANCE);
-         }
+      if (segments == null || segments.isEmpty()) return;
+
+      SearchMapping searchMapping = searchMappingHolder.getSearchMapping();
+      if (searchMapping == null) {
```
[the remark this answers is not in the document]
Comment (fax4ever) [reply query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java:341]: There was for the same reason of https://github.com/infinispan/infinispan/pull/8571#discussion_r458793497.
Going to remove these checks...
Moreover, probably thanks to https://github.com/infinispan/infinispan/pull/8575 we can get rid of the entire `SearchMappingHolder` class, but maybe this could be a follow up. What do you think about that? Thanks

**Comment by anistor** [comment]
java.lang.NullPointerException
	at org.infinispan.client.hotrod.query.RemoteQueryDisableIndexingTest.testEmptyIndexIsPresent(RemoteQueryDisableIndexingTest.java:113)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:124)
	at org.testng.internal.MethodInvocationHelper$1.runTestMethod(MethodInvocationHelper.java:230)

**Comment by fax4ever** [comment]
> java.lang.NullPointerException
> at org.infinispan.client.hotrod.query.RemoteQueryDisableIndexingTest.testEmptyIndexIsPresent(RemoteQueryDisableIndexingTest.java:113)
> at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> at java.base/java.lang.reflect.Method.invoke(Method.java:566)
> at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:124)
> at org.testng.internal.MethodInvocationHelper$1.runTestMethod(MethodInvocationHelper.java:230)

Commit https://github.com/infinispan/infinispan/pull/8571/commits/a65ee7e89be3ab5ffbb74a5d221da1a437f34e19 would have fixed it. The CI log says that there is no space on the device. See https://ci.infinispan.org/job/Infinispan/job/PR-8571/5/console :(

File: query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (fax4ever) [reply query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java:341]: Looking deeper at the issue, I think that we should keep these null checks or change the way we handle the case where we declare a cache as indexed but we don't provide any entity in the `indexedEntities` configuration.

Basically, if we remove these checks, we will have these three tests not working:
1. `ContinuousQueryObjectStorageTest`
2. `DataFormatIndexedTest`
3. `ClientListenerWithDslFilterObjectStorageTest`

In all these cases we configure the cache in this way:
``` java
cfgBuilder.indexing().enable()
            .addProperty("directory.type", "local-heap");
```

The cache is indexed but we don’t provide any indexed entity in the configuration.
I think that it would be the same if we defined some entities, but all of them were not contained in the serialization context at the time of cache starting.

Without the check we will get a null pointer exception locally, that will produce a timeout exception remotely.
The is issue is bound to the methods:
1. QueryInterceptor#purgeAllIndexes
2. QueryInterceptor#purgeIndex by segment

To be honest, I have no idea why a purge could be invoked on an indexed cache without no entities. Can you help me with that? Thanks

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/marshall/NonIndexedEmbeddedRemoteQueryTest.java [hunk 16]
```diff
@@ -25,10 +24,4 @@ protected ConfigurationBuilder createConfigBuilder() {
       builder.encoding().value().mediaType(MediaType.APPLICATION_OBJECT_TYPE);
       return builder;
    }
-
-   @Test(expectedExceptions = HotRodClientException.class, expectedExceptionsMessageRegExp = "org.infinispan.objectfilter.ParsingException: ISPN028521: Full-text queries cannot be applied to property 'longDescription' in type sample_bank_account.Transaction unless the property is indexed and analyzed.")
-   @Override
-   public void testRemoteFullTextQuery() {
-      super.testRemoteFullTextQuery();
-   }
```
[the remark this answers is not in the document]
Comment (fax4ever) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/marshall/NonIndexedEmbeddedRemoteQueryTest.java:33]: I removed the test since I had to change the base implementation, that is `EmbeddedRemoteInteropQueryTest#testRemoteFullTextQuery`.
The change is caused by the fact that Hibernate Search 6 does not support a field that is sortable and full text at the same time.

Anyway, I think that we can keep the original implementation for the subclass, so that we can keep the test case. I'm doing that...

File: core/src/main/java/org/infinispan/configuration/cache/IndexingConfigurationBuilder.java [hunk 17]
```diff
@@ -29,11 +29,12 @@
  */
 public class IndexingConfigurationBuilder extends AbstractConfigurationChildBuilder implements Builder<IndexingConfiguration>, ConfigurationBuilderInfo {
 
-   private static final String DIRECTORY_PROVIDER_SUFFIX = ".directory_provider";
+   private static final String DIRECTORY_PROVIDER_SUFFIX = "directory.type";
 
+   // TODO Check if *KEY1* can be used
    private static final String DIRECTORY_PROVIDER_KEY1 = "hibernate.search.default.directory_provider";
```
Comment (anistor) [inline core/src/main/java/org/infinispan/configuration/cache/IndexingConfigurationBuilder.java:35]: I suspect the suffix ```directory_provider``` here also must become ```directory.type```

File: query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (gustavonalle) [reply query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java:341]: Let's just handle it in a follow up

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/RemoteQueryNonQueryableCacheTest.java [hunk 18]
```diff
@@ -53,18 +54,22 @@ private Configuration createCache(String mediaType) {
...
       executeQuery(DEFAULT_CACHE);
-      executeQuery(INDEXED_CACHE);
       executeQuery(PROTOBUF_CACHE);
       executeQuery(POJO_CACHE);
    }
 
+   @Test(expectedExceptions = HotRodClientException.class, expectedExceptionsMessageRegExp = ".*ISPN014054.*")
+   public void assertErrorForIndexedCacheButWithoutIndexedEntities() {
+      executeQuery(INDEXED_CACHE);
+   }
```
[the remark this answers is not in the document]
Comment (fax4ever) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/RemoteQueryNonQueryableCacheTest.java:71]: I'll change the configuration to add an indexed protobuf entity, providing its protobuf schema in the serialization context. Doing now..

[the remark this answers is not in the document]
Comment (fax4ever) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/RemoteQueryNonQueryableCacheTest.java:71]: Since that now declaring a protobuf entity without its protobuf schema has no impact...

File: query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (anistor) [reply query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java:341]: Let's assume ```indexedEntities``` will always have at least 1 declared entity starting with ispn 12

File: core/src/main/java/org/infinispan/configuration/cache/IndexingConfigurationBuilder.java (hunk 17 above)
Comment (fax4ever) [reply core/src/main/java/org/infinispan/configuration/cache/IndexingConfigurationBuilder.java:35]: Doing the change now... Anyway the full property path is `hibernate.search.backends.infinispan_backend.directory.type`, see `org.infinispan.search.mapper.mapping.impl.IndexProperties`, in particular the line:

``` java
onfigurationPropertySource basePropertySource =
            propertyChecker.wrap(ConfigurationPropertySource.fromMap(backendProperties))
                  .withPrefix(BACKEND_PROPERTIES_PREFIX);
```

Of course you can arrange the simplification of the path as you wish, I did here an example of that.

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/MultiServerStoreQueryTest.java [hunk 19]
```diff
@@ -90,12 +91,13 @@ public Configuration buildIndexedConfig(String storeName) {
...
-             .addProperty("default.indexwriter.merge_factor", "30")
-             .addProperty("default.indexwriter.merge_max_size", "1024")
-             .addProperty("default.indexwriter.ram_buffer_size", "256");
+             .addProperty(SearchConfig.DIRECTORY_TYPE, SearchConfig.HEAP)
+            .addProperty(SearchConfig.THREAD_POOL_SIZE, "6")
+             .addProperty(SearchConfig.QUEUE_COUNT, "6")
+             .addProperty(SearchConfig.QUEUE_SIZE, "4096")
+             .addProperty(SearchConfig.COMMIT_INTERVAL, "10000")
+             .addProperty(SearchConfig.SHARDING_STRATEGY, SearchConfig.HASH)
+             .addProperty(SearchConfig.NUMBER_OF_SHARDS, "6");
```
[the remark this answers is not in the document]
Comment (fax4ever) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/MultiServerStoreQueryTest.java:100]: When I did the change we didn't have all the options we have right now. I'm restoring the old config using the new properties. This looks like the last comment I still have to fix. Isn’t it?

[the remark this answers is not in the document]
Comment (fax4ever) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/query/MultiServerStoreQueryTest.java:100]: Moreover, it looks like we don't need a constant class now, we can use the official one, that is "org.hibernate.search.backend.lucene.cfg.LuceneIndexSettings". Doing the change now...

**Comment by tristantarrant** [comment]
Ci looks ok to me aside from some minor possibly related failures. I'm going to merge it for the release tomorrow unless there are objections

**Comment by anistor** [comment]
no more objections from me, but @gustavonalle had some and I did not verify if the latest push fixed all of them. I'll put my suggested changes in a separate branch an PR after the release; no need to stall it

**Comment by tristantarrant** [comment]
Merged. Thanks for the big effort  @fax4ever @yrodiere

**Comment by fax4ever** [comment]
It was a great pleasure. We're really happy about this merge. Thanks to you guys ;)
